src/USER-MDI/fix_mdi_engine.cpp

```diff
@@ -0,0 +1,932 @@
+      error->all(FLERR,"Unable to send charges to driver");
+  }
+
+  delete [] charges;
+  delete [] charges_reduced;
+}
+
+
+void FixMDIEngine::send_energy(Error* error)
+{
+  double kelvin_to_hartree;
+  MDI_Conversion_Factor("kelvin_energy", "hartree", &kelvin_to_hartree);
+
+  double pe = potential_energy;
+  double ke = kinetic_energy;
+  double total_energy;
+  double *send_energy = &total_energy;
+
+  // convert the energy to atomic units
+  pe *= kelvin_to_hartree/force->boltz;
+  ke *= kelvin_to_hartree/force->boltz;
+  total_energy = pe + ke;
+
+  if (master) {
+    ierr = MDI_Send((char*) send_energy, 1, MDI_DOUBLE, driver_socket);
+    if (ierr != 0)
+      error->all(FLERR,"Unable to send potential energy to driver");
+  }
+}
+
+
+void FixMDIEngine::send_pe(Error* error)
+{
+  double kelvin_to_hartree;
+  MDI_Conversion_Factor("kelvin_energy", "hartree", &kelvin_to_hartree);
+
+  double pe = potential_energy;
+  double *send_energy = &pe;
+
+  // convert the energy to atomic units
+  pe *= kelvin_to_hartree/force->boltz;
+
+  if (master) {
+    ierr = MDI_Send((char*) send_energy, 1, MDI_DOUBLE, driver_socket);
+    if (ierr != 0)
+      error->all(FLERR,"Unable to send potential energy to driver");
+  }
+}
+
+
+void FixMDIEngine::send_ke(Error* error)
+{
+  double kelvin_to_hartree;
+  MDI_Conversion_Factor("kelvin_energy", "hartree", &kelvin_to_hartree);
+
+  double ke = kinetic_energy;
+  double *send_energy = &ke;
+
+  // convert the energy to atomic units
+  ke *= kelvin_to_hartree/force->boltz;
+
+  if (master) {
+    ierr = MDI_Send((char*) send_energy, 1, MDI_DOUBLE, driver_socket);
+    if (ierr != 0)
+      error->all(FLERR,"Unable to send potential energy to driver");
+  }
+}
+
+
+void FixMDIEngine::send_types(Error* error)
+{
+  int * const type = atom->type;
+
+  if (master) { 
+    ierr = MDI_Send((char*) type, atom->natoms, MDI_INT, driver_socket);
+    if (ierr != 0)
+      error->all(FLERR,"Unable to send atom types to driver");
+  }
+}
+
+
+void FixMDIEngine::send_labels(Error* error)
+{
+  char *labels = new char[atom->natoms * MDI_LABEL_LENGTH];
+  memset(labels, ' ', atom->natoms * MDI_LABEL_LENGTH);
+
+  for (int iatom=0; iatom < atom->natoms; iatom++) {
+    std::string label = std::to_string( atom->type[iatom] );
+    int label_len = std::min( int(label.length()), MDI_LABEL_LENGTH );
+    strncpy(&labels[iatom * MDI_LABEL_LENGTH], label.c_str(), label_len);
+  }
+
+  if (master) { 
+    ierr = MDI_Send( labels, atom->natoms * MDI_LABEL_LENGTH, MDI_CHAR, driver_socket);
+    if (ierr != 0)
+      error->all(FLERR,"Unable to send atom types to driver");
+  }
+
+  delete [] labels;
+}
+
+
+void FixMDIEngine::send_masses(Error* error)
+{
+  double * const mass = atom->mass;
+  int * const type = atom->type;
+
+  if (master) { 
+    double *mass_by_atom = new double[atom->natoms];
+    for (int iatom=0; iatom < atom->natoms; iatom++) {
+      mass_by_atom[ atom->tag[iatom] - 1 ] = mass[ type[iatom] ];
```

Comment:
This will fail with atom styles that use per-atom masses or when per-atom masses are enabled via fix property/atom (e.g. for simulation isotopes)  and thus this should be using `atom->rmass` instead of `atom->mass` to access masses.